src/engine.py

```
@@ -7145,6 +7193,31 @@ async def run_turn(transcript_text: str) -> None:
                         response_text = (str(llm_result) or "").strip()
                         tool_calls = []
 
+                    # Contexts are the source of truth for tool allowlisting: enforce at execution time too.
+                    allowed_tools: set[str] = set()
+                    try:
+                        allowed_tools = set((llm_options or {}).get("tools") or [])
+                    except Exception:
```

Ruff (0.14.10)

Do not catch blind exception: `Exception`

(BLE001)